### DISCUSSION TITLE
Persisted data removal on error

### DISCUSSION THREAD

**Discussion by salembaira** [TanStack/query#3623] [Q&A] [answered]
Hello guys and thank you for this awesome library.

I'm really not sure if this is a bug or not. Let me explain :

- when using `useQuery`, data is available after error state, assuming it was successful before
- when using `persistClient`, the query data is removed after an error occurs

I find these two behaviors inconsistent with one another. What if my API is having issues and I want the user to still get the old data even after reloading the app ? I could then, use the error state to warn the user that we are experiencing issues.

**Comment by karol-janik** [comment]
@TkDodo already created a great blog post, in which he present the idea to show the data first, maybe that will help, PTAL: https://tkdodo.eu/blog/status-checks-in-react-query

**Comment by TkDodo** [comment, marked as answer]
- `data` is never deleted due to a background fetch, even if it fails, as you might want to prefer showing stale data.
- the query will still be in `error` state though, and only queries that are in `success` state will be persisted. This can be customized by passing `shouldDehydrateQuery` to `dehydrateOptions`. The default is:
https://github.com/tannerlinsley/react-query/blob/5848fab8a560efcf66ef0062c207c3004bccad83/src/core/hydration.ts#L72-L74

but you can change that to e.g. `!!query.state.data` to include all queries that have data, even erroneous ones.

thinking about it, this might even be the "better" default implementation, we could change that for v4 🤔

**Reply by salembaira** [reply]
> data is never deleted due to a background fetch, even if it fails, as you might want to prefer showing stale data.

Sorry to say it really is deleted. Here is the proof using the offline example from v4 : 

This is a screenshot of the clientState after running the example with no issues. You can clearly see the movies list and the two details persisted as expected : 

![1](https://user-images.githubusercontent.com/5349435/169387540-c1495352-43d1-46c8-8c23-83fc490d9b13.PNG)

Now if we make `api.fetchMovies` return 403 and thus making ky return an error you can see that the `movies list` has disappeared from the cache : 

![2](https://user-images.githubusercontent.com/5349435/169388083-6a019993-51e0-4bdf-abbc-b7c603137c67.PNG)


When checking the documentation for v4 ([https://react-query-beta.tanstack.com/plugins/persistQueryClient](https://react-query-beta.tanstack.com/plugins/persistQueryClient)) it says : 

![3](https://user-images.githubusercontent.com/5349435/169388734-53b9d676-9e58-4a89-972a-cedddcb7dda5.PNG)

Is the third condition `error (ex: throws ...)` why this is happening ? If so, what is the reasoning behind this ? Why would you discard the cache in this case and not let the developer decide if he want to use it anyway ?

**Reply by TkDodo** [reply]
It is not deleted from the cache. You are showing examples of what gets persisted. Please read my comment again about dehydrateOptions. Per default, queries that are in error state are NOT persisted, but you can change that.

**Reply by salembaira** [reply]
Do you mean that if I did what you mentioned about dehydrateOptions the query will still be persisted even if it goes into error state ? If that is the case, this should definitely be the default behavior in my opinion.

**Reply by TkDodo** [reply]
I would mean that when dehydrating, the query would be persisted if it has data, independent of its state. If a query fails on the first try and thus doesn't have data, there's usually no point in persisting it.

**Reply by salembaira** [reply]
Thank you so much for explaining this. I hope this gets to be the default behavior in v4. It would be much more consistent with what happens in the cache. Or at least it should be the case when networkMode is “offlineFirst”.

**Reply by TkDodo** [reply]
we just discussed this, and the problem is that `Errors` can be serialized to json. so: `error: new Error("foo")` will become `error: {}`, so code can get brittle if you do `error.message.toUppercase()`, as it will work normally, but once restored from persistence, it will error at runtime.

with that in mind, we will not change the default for v4 because it would be unexpected.

**Reply by salembaira** [reply]
@TkDodo why not make something like this : (tested on a react-native app and works great)

```javascript
function serializeError(error) {
  if (!error) {
    return error;
  }
  const newError = {
    ...error,
    message: error.message
  };
  return newError;
}

function dehydrateQuery(query) {
  return {
    state: {
      ...query.state,
      error: serializeError(query.state.error)
    },
    queryKey: query.queryKey,
    queryHash: query.queryHash
  };
}
```
For a offlineFirst application this is really the expected behavior. This would be a missed opportunity if we don't include it in v4 as any offlineFirst app especially mobile apps expect data to stay persisted even if API throws an error at some point. This is alreay the default behaviour in the cache as the query state is updated but data is never deleted.

**Reply by TkDodo** [reply]
- that error will just be an object and not `instanceof Error`
- errors can be sub-classed (extended), like `axios` does with their `AxiosError`. All those things would get lost.

**Reply by salembaira** [reply]
So we are gonna let the whole state get lost just because we can’t stringify the error instance ? What if we add something like `isPersistedError` to signal that this is a simple error object.

Anyway, I won’t complain too much as my current patch for `react-query` does the job but it would be really nice if this could be something that the library handles by default or at least is configurable.

**Reply by TkDodo** [reply]
This is just the default behaviour. You can customize this for your needs by:
- setting `shouldDehydrateQueries` in `dehydrateOptions` to your needs
- you can also set a custom `serialize` function to serialize errors the way you want. You can also use something like [superjson](https://github.com/blitz-js/superjson) if you want

most people likely just show the error if the query is in error state, something like:

```
if (query.isError) {
  return 'Error:' + error.message
}
```

if you ignore background errors and show stale data instead _and_ want to persist that, you can do so, but the default behaviour won't change for the reasons I've mentioned. It's literally 2 lines of code to setup on the global persistence plugin :)

**Reply by salembaira** [reply]
You are right. A custom serialize function + shouldDehydrateQueries seems more than enough. Do you think this is worth mentioning in the docs ? I suspect a lot of offline-first apps need this.

**Reply by TkDodo** [reply]
Sure, docs can always be improved. Feel free to add a PR to the beta branch
